ta/pkcs11/src/token_capabilities.c

```diff
@@ -152,75 +121,51 @@ bool mechanism_flags_complies_pkcs11(uint32_t mechanism_type, uint32_t flags)
 	return false;
 }
 
-/*
- * struct supported_mechanism - Describe mechanism supported by the PKCS11 TA
- * @id: ID of the mechanism
- * @flags: PKCS11_CKFM_* flags supported tby the TA for the mechanism
- * @string: Helper string of a mechanism ID for debug purpose
- */
-struct supported_mechanism {
-	enum pkcs11_mechanism_id id;
-	uint32_t flags;
-#if CFG_TEE_TA_LOG_LEVEL > 0
-	const char *string;
-#endif
-};
-
-#if CFG_TEE_TA_LOG_LEVEL > 0
-#define TOKEN_MECHANISM(_proc_id, _supported_flags)		\
-	{							\
-		.id = _proc_id,					\
-		.flags = (_supported_flags),			\
-		.string = #_proc_id,				\
-	}
-#else
-#define TOKEN_MECHANISM(_proc_id, _supported_flags)		\
-	{							\
-		.id = _proc_id,					\
-		.flags = (_supported_flags),			\
-	}
-#endif
-
 /*
  * Arrays that centralizes the IDs and processing flags for mechanisms
  * supported by each embedded token.
  */
-const struct supported_mechanism token_mechanism[] = {
-	TOKEN_MECHANISM(PKCS11_CKM_AES_ECB, 0),
+const struct pkcs11_mechachism_modes token_mechanism[] = {
+	MECHANISM(PKCS11_CKM_AES_ECB,
+		  PKCS11_CKFM_ENCRYPT | PKCS11_CKFM_DECRYPT, 0),
 };
 
 /*
- * get_token_supported_mechanisms - Get the mechasnism supproted by the token
- * @array: Output array to be filled with mechnism IDs (32bit IDs) or NULL
- * @count: Number of 32bit IDs allocated for @array if @array is not NULL
- * Return the number of mechanisms supported by the token
+ * tee_malloc_mechanism_array - Allocate and fill array of supported mechanisms
+ * @count: [in] [out] Pointer to number of mechanism IDs in client resource
+ * Return allocated array of the supported mechanism IDs
+ *
+ * Allocates array with 32bit cells mechanism IDs for the supported ones only
+ * if *@count covers number mechanism IDs exposed.
  */
-size_t get_supported_mechanisms(uint32_t *array, size_t array_count)
+uint32_t *tee_malloc_mechanism_list(size_t *out_count)
```

Comment:
This wasn't really what I meant. I guess it works too, however you're going to have to deal with the freeing at the callers end. What I had in mind was an interface where the supplied buffer was filled in if large enough with the required size reported in all occasions. Do what you prefer.

Reply:
> (...) you're going to have to deal with the freeing at the callers end (...)

This mechanism array is only called upon `GetMechanismList()`, so related TA cmd handler simply calls `TEE_Free()` on the reference once done with its content. Note array is allocated only upon client buffer size, __line 151__ `if (*out_count >= count)`.


> where the supplied buffer was filled in if large enough with the required size reported in all occasions

The mechanism list is currently built at runtime in an allocated u32 array that in `TEE_MemMove()` in one shoot in client buffer.
An alternate is to `TEE_MemMove()` __each__ u32 mechanism IDs straight to client buffer.